### PR TITLE
Disallow multiple commands in a single CommandRequest

### PR DESCRIPTION
**Pull request by otterdahl**
Fixes issue: https://github.com/rsmp-nordic/rsmp_core/issues/169

### CODE REVIEW

**Comment by emiltin** [comment]
Is sending multiple commands in a single message not used in practice?
Strictly speaking, this is breaking change.

**Comment by otterdahl** [comment]
I don't believe it is used in practice. But I might be mistaken.

I believe that commands are fairly seldomly sent compared to other message types, for instance alarms or statuses. Therefore there is little point for supervision systems to try optimize network usage by grouping multiple commands together in a single CommandRequest. I have never seen any system trying to do this.

But strictly speaking you're right. It might be a breaking change.

**Comment by marcgarba** [comment]
In the first paragraph of the current specification of command :
"Command messages are used to give order using _one or more_ commands, for the referenced object. The
site responds with a command acknowledgement."

why still evoking "one or more commands" ? or perhaps here it is '**arguments**/**names**' of the same command (code Id).
 
 And also later :
 "The command code (cCI) and **name** (n) are placed in an array (arg) in order to enable support for
requesting multiple commands at once."

I think there is an ambiguity about general "command" term, if corresponding to code id part, or name part !
And also some time we talk of  "argument" and some time "name", instead of 'argument name'...?

---

Mistake in the commit for error handling part:
"If more than one command is included in a single CommandRequest, _than_ this is
considered a serious error resulting in MessageNotAck." => **then** this is...

**Comment by otterdahl** [comment]
Thank you for the feedback. I've updated the PR.

> In the first paragraph of the current specification of command : "Command messages are used to give order using _one or more_ commands, for the referenced object. The site responds with a command acknowledgement."
> 
> why still evoking "one or more commands" ?

You're right. It should only be a single command.

> And also later : "The command code (cCI) and **name** (n) are placed in an array (arg) in order to enable support for requesting multiple commands at once."

Fixed

> I think there is an ambiguity about general "command" term, if corresponding to code id part, or name part ! And also some time we talk of "argument" and some time "name", instead of 'argument name'...?

The intention is that a "Command" refers to a single command code id ("cCI"). Each argument has it's own "name", "command" and associated value.

> Mistake in the commit for error handling part: "If more than one command is included in a single CommandRequest, _than_ this is considered a serious error resulting in MessageNotAck." => **then** this is...

Yes. It's a typo. Fixed.

**Comment by marcgarba** [comment]
For "The command code (``cCI``) and name (``n``) are placed in an array (``arg``)."
Proposal change : "The command code (``cCI``) and name **of argument** (``n``) are placed in an array (``arg``)."
(corresponding to description of element ``n`` in table 4.19)

File: source/applicability/basic_structure.rst [hunk 1]
```diff
@@ -1187,13 +1187,16 @@ Example of message exchange with subscription, status updates and unsubscription
...
 
-Command messages are used to give order using one or more commands, for the
-referenced object.
+Command messages are used to give order to the referenced object.
 The site responds with a command acknowledgement.
 
 All arguments needs to included in a command, otherwise it results a serious
 error resulting in MessageNotAck. See section about :ref:`incomplete-commands`.
 
+Only a single command is allowed in single CommandRequest, otherwise any
```
Comment (emiltin) [inline source/applicability/basic_structure.rst]: => Only a single command is allowed in a CommandRequest, otherwise any

Comment (otterdahl) [reply source/applicability/basic_structure.rst]: Fixed

**Comment by otterdahl** [comment]
> For "The command code (`cCI`) and name (`n`) are placed in an array (`arg`)." Proposal change : "The command code (`cCI`) and name **of argument** (`n`) are placed in an array (`arg`)." (corresponding to description of element `n` in table 4.19)

Thank you for the input. I've updated the PR with a3d7e3d4456250791d34eea22b4c855787c0ad2c.

**Comment by emiltin** [comment]
I suggest adding a note somewhere that command responses can only have one command code item, otherwise it's an invalid message.

**Comment by otterdahl** [comment]
> I suggest adding a note somewhere that command responses can only have one command code item, otherwise it's an invalid message.

I agree. PR updated.

**Comment by marcgarba** [comment]
This PR has been closed, but for the "Structure of a command **response** message" part, a sentence has not been modified to take into acount the "single" command constraint.

Actually : 

_"The command code (cCI) and name (n) are placed in an array (rvs) in order to enable support for responding to multiple commands at once."_

Suggested : 

_"The command code (cCI) and name of the argument (n) are placed in an array (rvs)."_

**Comment by emiltin** [comment]
> This PR has been closed, but for the "Structure of a command **response** message" part, a sentence has not been modified to take into acount the "single" command constraint.

Opened #241 for this.

**Pull request merged**